### DISCUSSION TITLE
Looking for trajectory optimization and PWA advice

### DISCUSSION THREAD

**Discussion by kirkrudolph** [yalmip/YALMIP#1281] [Optimization modelling] [answered]
Hello,

I'm trying to plan an optimal trajectory for my constrained MIMO control system. The end goal would be a ~3 minute, optimal, open-loop trajectory that I'd be able to track with an online controller.

My system:
1. has 3 continuous states (`x`), 3 actuators (`u`), 2 measured disturbances (`d`), 6 sensors (`y`), and 21 outputs (`yhat`)
   - Note: The system is fully observable and controllable
2. is currently discretized at 5ms
   - With a goal of a 3 minute trajectory, that's a horizon (`N`) of ~36,000. I have no idea if that's feasible.
   - Other discrete rates / continuous dynamics / descriptor state-space could be used if it'd help reduce the size of the problem and/or take advantage of sparsity.
3. has 9 "modes"
   - The mode itself is technically an integer state (for example, 1->2, 1<-2->3, 2<-3->4, etc.)
   - The transition between modes is determined from the first two states and the current mode.
   - The states `x` are continuous across mode changes but the actuation `u` is usually discontinuous
   - The different modes each have unique linear dynamics (`A`,`B`,`E`,`C`, etc.)
   - Is this considered piecewise affine (PWA)?
5. Quadratic (or possibly linear) objective
   - Looking for "minimum fuel" trajectories (subject to constraints)
6. Many linear equality and inequalities on both the states and actuators.
7. Bilinear equalities and inequalities (however, I haven't added these yet to keep the problem as simple as possible for the time being)

Looking through the YALMIP documentation, I see many different ways to approach this problem class but am struggling to understand which best fits my goal and system:
1. Combination of [Modeling if else](https://yalmip.github.io/example/modellingif/) + [Hybrid MPC](https://yalmip.github.io/example/hybridmpc/)
2. [Explicit MPC](https://yalmip.github.io/example/explicitmpc/)
3. making a fourth state an `intvar` for the current "mode"

Any advice on how to approach this problem would be greatly appreciated!

Thank you,
Kirk Rudolph

**Comment by johanlofberg** [comment, marked as answer]
Had it not been for the last item it is just standard hybrid MPC (i.e standard MILP modelling)

However, when you add bilinear stuff, it goes from very hard to extremely hard most likely intractable. I would say there is no reasonable way to attack that in YALMIP still expecting sensible performance/optimal solutions (assuming the bilinear stuff isn't trivially linearizable i.e. products between continuous and binary, or that you mean convex quadratic constraints)

No reason to talk about explicit MPC, as you only want an open-loop solution and thus has no need for massive computations to compute a closed-loop operator.

**Reply by kirkrudolph** [reply]
Johan, 

Thanks for the quick reply. I called it a bilinear constraint because that's how it showed up in the `lmi` constraint object. The inequality constraints are actuator power limits. For example,
```
 % bilinear constraints
 PowerLimits = [    0 <= x{k}(1)*u{k}(1) <= 200,... 
                 -200 <= x{k}(2)*u{k}(2) <= 200,...
                 -200 <= x{k}(1)*u{k}(3) <= 200];
```
Results in
```
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
|    ID|                               Constraint|   Coefficient range|
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
|   #14|   Element-wise inequality (bilinear) 2x1|            1 to 200|
|   #15|   Element-wise inequality (bilinear) 2x1|            1 to 200|
|   #16|   Element-wise inequality (bilinear) 2x1|            1 to 200|
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

Also, the nonlinear dynamics which accumulates power
```
[ x{k+1}(3) == x{k}(3) + x{k}(2)*u{k}(2) + x{k}(1)*u{k}(3) ]
```

I'm not an expert at identifying the different classes of problem yet so I'm not sure if these these specific nonlinear constraints are "convex quadratic constraints" like you mentioned above. I feel like they might be?

Thanks,
Kirk Rudolph

**Reply by johanlofberg** [reply]
if both of those variables in the bilinear constraints are continuous, you're simply not going to solve that problem without significant effort  in developing specialized solvers, or money to buy solvers and hope they can do it

bilinear integer models with tens of variables can be challenging, hundreds would be something that would make developers of state-of-the-art solvers proud, and thousands is what you brag about when show-casing recent research

**Reply by kirkrudolph** [reply]
Thank you for your insight! For my system, I believe I should be able to get pretty far by simply ignoring those addition problematic constraints or by being overly restrictive. Either way, you've probably saved me a ton of time! Thank you so much and great work on YALMIP!

**Comment by johanlofberg** [comment]
and an horizon of 36000 is just ridiculously out of scope for pwa/hybrid-mpc stuff (and indicates waaaaay too fast sampling. That's like planning a day ahead using a sampling-rate of 2.4 seconds, which makes no sense)

**Reply by kirkrudolph** [reply]
I kind of figured that size of problem was unreasonable. Thank you for the confirmation. The 3-minute trajectory is the goal of this optimization. The only reason I'm using a 5ms discretization is because I already had it handy from an online controller development. The online controller usually requires a step change in actuator effort after a mode change. If the mode switch occurs and the actuators hold the previous value, the system does the exact opposite of what's desired and usually induces mode switching. 

I'll plan on discretizing the continuous time dynamics at a much slower rate unless you have other suggestions. Thanks again!

**Reply by johanlofberg** [reply]
yes, separate low-level control sample-rate from planning
